src/sentry/utils/kvstore/bigtable.py

```diff
@@ -0,0 +1,287 @@
+import enum
+import logging
+import struct
+import zlib
+from datetime import timedelta
+from threading import Lock
+from typing import Any, Callable, Iterator, Mapping, NamedTuple, Optional, Sequence, Tuple, cast
+
+import zstandard
+from django.utils import timezone
+from google.api_core import exceptions, retry
+from google.cloud import bigtable
+from google.cloud.bigtable.row_data import PartialRowData
+from google.cloud.bigtable.row_set import RowSet
+from google.cloud.bigtable.table import Table
+from sentry.utils.kvstore.abstract import KVStorage
+
+
+logger = logging.getLogger(__name__)
+
+
+class CompressionStrategy(NamedTuple):
+    compress: Callable[[bytes], bytes]
+    decompress: Callable[[bytes], bytes]
+
+
+class BigtableError(Exception):
+    pass
+
+
+class BigtableKVStorage(KVStorage[str, bytes]):
+    max_size = 1024 * 1024 * 10
+    column_family = "x"
+
+    ttl_column = b"t"
+    ttl_struct = struct.Struct("<I")
+    flags_column = b"f"
+    flags_struct = struct.Struct("B")
+    data_column = b"0"
+
+    class Flags(enum.IntFlag):
+        # XXX: Compression flags are assumed to be mutually exclusive, the
+        # behavior is explicitly undefined if both bits are set on a record.
+        COMPRESSED_ZLIB = 1 << 0
+        COMPRESSED_ZSTD = 1 << 1
+
+    compression_strategies: Mapping[str, Tuple[Flags, CompressionStrategy]] = {
+        "zlib": (
+            Flags.COMPRESSED_ZLIB,
+            CompressionStrategy(compress=zlib.compress, decompress=zlib.decompress),
+        ),
+        "zstd": (
+            Flags.COMPRESSED_ZSTD,
+            CompressionStrategy(
+                compress=lambda x: cast(bytes, zstandard.ZstdCompressor().compress(x)),
+                decompress=lambda x: cast(bytes, zstandard.ZstdDecompressor().decompress(x)),
+            ),
+        ),
+    }
+
+    def __init__(
+        self,
+        project: Optional[str] = None,
+        instance: str = "sentry",
+        table_name: str = "nodestore",
+        client_options: Optional[Mapping[Any, Any]] = None,
+        default_ttl: Optional[timedelta] = None,
+        compression: Optional[str] = None,
+    ) -> None:
+        client_options = client_options if client_options is not None else {}
+        if "admin" in client_options:
+            raise ValueError('"admin" cannot be provided as a client option')
+
+        if compression is not None and compression not in self.compression_strategies:
+            raise ValueError(f'"compression" must be one of {self.compression_strategies.keys()!r}')
+
+        self.project = project
+        self.instance = instance
+        self.table_name = table_name
+        self.client_options = client_options
+        self.default_ttl = default_ttl
+        self.compression = compression
+
+        self.__table: Table
+        self.__table_lock = Lock()
```

Comment:
This was brought in as an instance variable (from the global `_connection_cache`) because I can't think of a situation where it'd be useful to have two separate `BigtableKVStorage` instances referring to the same `Table`.

I do think there are cases where it might be useful for multiple instances to share the same `Client`. As far as I can tell, multiple `Table` instances can do that as `Instance` and `Table` instances maintain references back to the client that created them, which holds all of the actual connection logic. I haven't looked too closely though, so I may have read that wrong.